ipn/ipnlocal/local.go

```diff
@@ -2014,6 +2011,16 @@ func (b *LocalBackend) WatchNotifications(ctx context.Context, mask ipn.NotifyWa
 		go b.pollRequestEngineStatus(ctx)
 	}
 
+	if mask&ipn.NotifyServeRequest != 0 {
+		defer func() {
+			sc := b.ServeConfig().AsStruct()
+			if sc != nil {
+				delete(sc.Foreground, sessionID)
+				b.SetServeConfig(sc) // TODO(marwan-at-work): check err
```

Comment:
this seems racy

instead add a method next to:

```
func (b *LocalBackend) SetServeConfig(config *ipn.ServeConfig) error {
	b.mu.Lock()
	defer b.mu.Unlock()
```

like `(*LocalBackend) DeleteForegroundServeConfig(sessionID string)`

That does the whole delete & conditional write under the same lock.